server/channels/api4/outgoing_oauth_connection.go

```diff
@@ -20,7 +21,35 @@ const (
 
 func (api *API) InitOutgoingOAuthConnection() {
 	api.BaseRoutes.OutgoingOAuthConnections.Handle("", api.APISessionRequired(listOutgoingOAuthConnections)).Methods("GET")
+	api.BaseRoutes.OutgoingOAuthConnections.Handle("", api.APISessionRequired(createOutgoingOAuthConnection)).Methods("POST")
 	api.BaseRoutes.OutgoingOAuthConnection.Handle("", api.APISessionRequired(getOutgoingOAuthConnection)).Methods("GET")
+	api.BaseRoutes.OutgoingOAuthConnection.Handle("", api.APISessionRequired(updateOutgoingOAuthConnection)).Methods("PUT")
+	api.BaseRoutes.OutgoingOAuthConnection.Handle("", api.APISessionRequired(deleteOutgoingOAuthConnection)).Methods("DELETE")
+}
+
+// checkOutgoingOAuthConnectionReadPermissions checks if the user has the permissions to read outgoing oauth connections.
+// The user needs to have the permissions to manage outgoing webhooks and slash commands in order to read outgoing oauth connections
+// so that they can use them in their outgoing webhooks and slash commands.
+// This is made in this way so only users with the management permission can setup the outgoing oauth connections and then
+// other users can use them in their outgoing webhooks and slash commands if they have permissions to manage those.
+func checkOutgoingOAuthConnectionReadPermissions(c *Context) bool {
+	if c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionManageOutgoingWebhooks) && c.App.SessionHasPermissionTo(*c.AppContext.Session(), model.PermissionManageSlashCommands) {
+		return true
+	}
+
+	c.SetPermissionError(model.PermissionManageOutgoingWebhooks, model.PermissionManageSlashCommands)
+	return false
+}
+
+// checkOutgoingOAuthConnectionWritePermissions checks if the user has the permissions to write outgoing oauth connections.
+// This is a more granular permissions intended for system admins to manage (setup) outgoing oauth connections.
+func checkOutgoingOAuthConnectionWritePermissions(c *Context) bool {
```

Comment:
I don't think that requiring additional permissions to read the outgoing oauth connections makes sense.

We require `PermissionManageOutgoingWebhooks` for write and `PermissionManageOutgoingWebhooks` AND `PermissionManageSlashCommands` for read. 

Apart from causing UI issues (see: https://github.com/mattermost/mattermost/pull/25507#issuecomment-1931934471), a member with `PermissionManageOutgoingWebhooks` but without `PermissionManageSlashCommands` can still read the outgoing connections. 

All they need to do is find the ID of the connection and send a PUT request to update it:

```
PUT /api/v4/oauth/outgoing_connections/wc5biebj4jgotdyp73fxgd9kbh

{"name":"aasaa","id":"wc5biebj4jgotdyp73fxgd9kbh"}
```

The response contains all connection details:
```json
{
"id":"wc5biebj4jgotdyp73fxgd9kbh",
"creator_id":"oytau6sje7rgfm7wysberta5qe",
"create_at":1707305608228,
"update_at":1707311494908,
"name":"aasaa",
"client_id":"example_client_id",
"oauth_token_url":"http://localhost:8065/plugins/oauth-provider-example/token",
"grant_type":"client_credentials",
"audiences":["https://mydomain.com/api/sq","https://mydomain.com/api/*"]
}

```
 
Based on the above, I'd suggest requiring only `PermissionManageOutgoingWebhooks` for both write and read operations.

Reply:
> The permissions required are as follow:

> WRITE: (Can CRUD outoing oauth connections): PermissionManageOutgoingOauthConnections
> READ: (Can read/list outgoing oauth connections): PermissionManageOutgoingWebhooks || PermissionManageSlashCommands

I mixed up the permissions. You're right about the READ and the reasoning for having them different makes sense. 

But currently the `PermissionManageOutgoingOauthConnections` doesn't give you Read access to the outgoing   oauth connections. It only gives you CUD, not R. 

We need to grand read access to Outgoing OAuth connection also to those with `PermissionManageOutgoingOauthConnections` because by having Update permissions they actually can read them too.

Otherwise we'd get reports that while someone with `PermissionManageOutgoingOauthConnections` can't read the connections, they can actually get their details but e.g. updating their name by only knowing the ID.

Reply:
You are totally right, let me fix that.

Reply:
Made the changes directly in the feature branch to avoid having to merge all previous PRs and changes into it again:  https://github.com/mattermost/mattermost/pull/25507/commits/ba46a59a8097c1516f28b906560ccc50a23ac828

Reply:
Added one more comment here: https://github.com/mattermost/mattermost/pull/25507/files#r1482819971. It's not about security. its about user expectations and experience.